File: Java/Strings/StringReversal.java
/*
Approach:
  - We first check if the input string is null or empty, in which case we return the input as it is.
  - Next, we convert the input string to a character array to make it easier to reverse.
  - We then initialize two pointers, `start` and `end`, pointing to the start and end of the character array, respectively.
  - Using a while loop, we swap the characters at `start` and `end` positions and increment `start` and decrement `end` until `start` becomes greater than or equal to `end`.
  - Finally, we convert the reversed character array back to a string using the `String` constructor and return it.

Time Complexity: The time complexity of this approach is O(n), where n is the length of the input string. This is because we need to iterate through the entire string to reverse it.

Space Complexity: The space complexity is O(n) as well. In this approach, we are using a character array to store the reversed string. The size of the character array is the same as the length of the input string.

Sample Input and Output:
  - Input: "Hello, World!"
  - Output: "!dlroW ,olleH"
*/

// here's a code in Java to reverse a string:

public class StringReversal {
    public static String reverseString(String input) {
        // Check if the string is null or empty
        if (input == null || input.isEmpty()) {
            return input;
        }
        
        // Convert the string to a character array
        char[] charArray = input.toCharArray();
        
        // Initialize pointers for the start and end of the array
        int start = 0;
        int end = charArray.length - 1;
        
        // Reverse the string by swapping characters from start to end
        while (start < end) {
            char temp = charArray[start];
            charArray[start] = charArray[end];
            charArray[end] = temp;
            
            start++;
            end--;
        }
        
        // Convert the character array back to a string
        return new String(charArray);
    }
    
    public static void main(String[] args) {
        // Testing the reverseString() method
        String input = "Hello, World!";
        String reversedString = reverseString(input);
        
        System.out.println("Original String: " + input);
        System.out.println("Reversed String: " + reversedString);
    }
}

File: C++/Strings/StringReversal.cpp
/*
Explanation:
- We define a function `reverseString` that takes a reference to a string as input. The function uses two pointers, `start` and `end`, initialized to the first and last characters of the string, respectively.
- We use a `while` loop that runs until the `start` pointer is less than the `end` pointer. In each iteration, we swap the characters at the `start` and `end` positions and then move both pointers inward towards the center of the string.
- The main function contains a sample input string, prints the original string, calls the `reverseString` function, and then prints the reversed string.

Sample Input:
"Hello, World!"

Sample Output:
Original String: Hello, World!
Reversed String: !dlroW ,olleH

Time Complexity: The `reverseString` function runs in linear time O(N), where N is the length of the string. This is because we traverse half of the string to perform the character swapping.
Space Complexity: The space complexity is O(1) since the algorithm only uses a constant amount of extra memory for the `start`, `end`, and `temp` variables, regardless of the string length.
*/

#include <iostream>
#include <string>

// Function to reverse the string
void reverseString(std::string& str) {
    int start = 0;
    int end = str.length() - 1;

    while (start < end) {
        // Swap characters at start and end positions
        std::swap(str[start], str[end]);

        // Move the start and end pointers inward
        start++;
        end--;
    }
}

int main() {
    // Sample input
    std::string str = "Hello, World!";

    // Print the original string
    std::cout << "Original String: " << str << std::endl;

    // Reverse the string
    reverseString(str);

    // Print the reversed string
    std::cout << "Reversed String: " << str << std::endl;

    return 0;
}
